### PR TITLE
Functionality to accept compressed files as input to predict when using a Predictor

### LINKED ISSUE #5237: Accept compressed files as input to `predict` when using a `Predictor`
**Is your feature request related to a problem? Please describe.**
I typically used compressed datasets (e.g. gzipped) to save disk space. This works fine with AllenNLP during training because I can write my dataset reader to load the compressed data. However, the `predict` command opens the file and reads lines for the `Predictor`. This fails when it tries to load data from my compressed files.

https://github.com/allenai/allennlp/blob/39d7e5ae06551fe371d3e16f4d93162e55ec5dcc/allennlp/commands/predict.py#L208-L218

**Describe the solution you'd like**
Either automatically detect the file is compressed or add a flag to `predict` that indicates that the file is compressed. One method that I have used to detect if a file is gzipped is [here](https://stackoverflow.com/questions/3703276/how-to-tell-if-a-file-is-gzip-compressed), although it isn't 100% accurate. I have an implementation [here](https://github.com/danieldeutsch/sacrerouge/blob/master/sacrerouge/io/util.py). Otherwise a flag like `--compression-type` to mark how the file is compressed should be sufficient. Passing the type of compression would allow support for gzip, bz2, or any other method.

### PR DESCRIPTION
**Pull request by Dbhasin1**
Fixes #5237 

Changes proposed in this pull request:

- Uses the open_compressed helper function in [allennlp/allennlp/common/file_utils.py](url) to handle compressed files in [allennlp/allennlp/commands/predict.py](url) and if our method to detect the compression format fails, asks the user to manually specify the format type as an argument to the predict class (if they haven't already stated so)
- Apart from gz and bz2 formats, added provision to deal with the lzma file format

## Before submitting

- [ ✅] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.
- [ ] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

File: allennlp/commands/predict.py [hunk 1]
```diff
@@ -212,10 +221,21 @@ def _get_json_data(self) -> Iterator[JsonDict]:
...
+            except OSError:
+                if self.compression_type:
+                    with open_compressed(input_file, self.compression_type) as file_input:
+                        for line in file_input:
+                            if not line.isspace():
+                                yield self._predictor.load_line(line)
+                else:
+                    print(
+                        "Automatic detection of compression type failed, please specify the compression type argument"
+                    )
```
Comment (dirkgr) [inline allennlp/commands/predict.py:238]: This logic needs to be the other way around. If the compression type is specified, we have to always respect it. If it's not specified, we autodetect.

File: allennlp/common/file_utils.py [hunk 2]
```diff
@@ -1085,20 +1085,27 @@ def get_file_extension(path: str, dot=True, lower: bool = True):
 
 
 def open_compressed(
-    filename: Union[str, PathLike], mode: str = "rt", encoding: Optional[str] = "UTF-8", **kwargs
+    filename: Union[str, PathLike],
+    compression_type: str = None,
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:1089]: This should be the last parameter, so we don't break existing usage of positional arguments.

Comment (dirkgr) [reply allennlp/common/file_utils.py:1089]: Also, the type of this should be `Optional[None]`.

File: allennlp/common/file_utils.py [hunk 3]
```diff
@@ -1085,20 +1085,27 @@ def get_file_extension(path: str, dot=True, lower: bool = True):
...
     open_fn: Callable = open
 
-    if filename.endswith(".gz"):
-        import gzip
-
-        open_fn = gzip.open
-    elif filename.endswith(".bz2"):
-        import bz2
+    compression_modules = {"gz": "gzip", "bz2": "bz2", "lzma": "lzma"}
+    if not compression_type:
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:1099]: If I pass in an empty string for `compression_type`, we will go down this path. I don't think that's what we want.

File: allennlp/common/file_utils.py [hunk 4]
```diff
@@ -1085,20 +1085,27 @@ def get_file_extension(path: str, dot=True, lower: bool = True):
...
+        for extension in compression_modules:
+            if filename.endswith(extension):
+                module = __import__(compression_modules[extension])
+                open_fn = module.open
+                break 
+    else:
+        module = __import__(compression_modules[extension])
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:1106]: I think `extension` is undefined here? Or am I blind and can't see it?

**Review by dirkgr — CHANGES_REQUESTED** [review]
Looks good, but needs a few tweaks.

I would be OK with not having the manual override functionality at all. If you name your gzip file `"something.bz2"`, you deserve what you get. But if you have a scenario in mind where we need this, let's keep it. It's good practice.

File: allennlp/commands/predict.py (hunk 1 above)
Comment (Dbhasin1) [reply allennlp/commands/predict.py:238]: That makes sense, will incorporate it

File: allennlp/common/file_utils.py (hunk 3 above)
Comment (Dbhasin1) [reply allennlp/common/file_utils.py:1099]: Can I simply consider an empty string for compression_type to be equivalent to it being 'None' ?

File: allennlp/common/file_utils.py (hunk 4 above)
Comment (Dbhasin1) [reply allennlp/common/file_utils.py:1106]: Yep, that was an error at my end, will fix it

File: allennlp/common/file_utils.py (hunk 3 above)
Comment (dirkgr) [reply allennlp/common/file_utils.py:1099]: I think that would be surprising to a user of the library. You can just compare to `None`:
```suggestion
    if compression_type is None:
```

File: allennlp/common/file_utils.py (hunk 2 above)
Comment (Dbhasin1) [reply allennlp/common/file_utils.py]: the `make typecheck` commands gave me incompatibility error on changing it to `Optional[None]` so I kept it as `Optional[str]`, will that be fine?

File: allennlp/commands/predict.py [hunk 5]
```diff
@@ -152,6 +160,7 @@ def __init__(
         batch_size: int,
         print_to_console: bool,
         has_dataset_reader: bool,
+        compression_type: str = None,
```
Comment (dirkgr) [inline allennlp/commands/predict.py:163]: The type should be `Optional[str]`.

File: allennlp/commands/predict.py [hunk 6]
```diff
@@ -212,10 +221,25 @@ def _get_json_data(self) -> Iterator[JsonDict]:
...
+                    )
+
+            else:
+                try:
+                    with open_compressed(input_file, compression_type=self.compression_type) as file_input:
+                        for line in file_input:
+                            if not line.isspace():
+                                yield self._predictor.load_line(line)
+                except OSError:
+                    print("please specify the correct compression type argument.")
```
Comment (dirkgr) [inline allennlp/commands/predict.py:242]: Why `OSError`? I don't think you have to catch this exception at all. If it fails, it fails. The only thing we might want to do is make sure that `open_compressed()` throws exceptions that are understandable.

File: allennlp/common/file_utils.py [hunk 7]
```diff
@@ -1085,20 +1085,27 @@ def get_file_extension(path: str, dot=True, lower: bool = True):
...
-        open_fn = gzip.open
-    elif filename.endswith(".bz2"):
-        import bz2
+    compression_modules = {"gz": "gzip", "bz2": "bz2", "lzma": "lzma"}
+    if compression_type in compression_modules:
+        module = __import__(compression_modules[compression_type])
+        open_fn = module.open
+    else:
+        for extension in compression_modules:
+            if filename.endswith(extension):
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:1104]: Can you use `os.path.splitext()` here to make that detection? I don't want a file named [`info.fogbugz`](https://fogbugz.com) to show up as a gzip file.

**Review by dirkgr — CHANGES_REQUESTED** [review]
Close to done!

**Comment by dirkgr** [comment]
Since I can't push to your branch and you are MIA, I'm continuing this in #5578.

**Pull request closed without merging**